pkg/registry/core/service/strategy.go

```diff
@@ -242,22 +247,15 @@ func normalizeClusterIPs(oldSvc *api.Service, newSvc *api.Service) {
 			return
 		}
 
-		// User specified plural and not singular (new client) so init singular for them.
-		if len(newSvc.Spec.ClusterIPs) > 0 && len(newSvc.Spec.ClusterIP) == 0 {
-			newSvc.Spec.ClusterIP = newSvc.Spec.ClusterIPs[0]
-			return
-		}
+		// we don't init singular based on plural because
```

Comment:
can we have mismatches like in https://github.com/kubernetes/kubernetes/pull/88505 with ClusterIPs?

Reply:
@thockin this is an important one. with PodIPs a fix was added after the fact to ensure if two are used and PodIP != PodIPs[0] then PodIP (old client) wins. I do believe this is a patching case. do you mind if added that to update case?

Reply:
on a second thought because of the special immutability that ClusterIP has we should be fine. because we handle clearing ClusterIPs case (e.g. old client PUTing/replacing with no ClusterIPs).

Reply:
Right, this handles that below, I think:


if clusterIPs was wiped, assume an old client just didn't know about it:

```
	if len(oldSvc.Spec.ClusterIPs) > 0 && len(newSvc.Spec.ClusterIPs) == 0 {
		// if ClusterIP is the same, then it is an old client trying to
		// patch service and didn't provide ClusterIPs
		if oldSvc.Spec.ClusterIP == newSvc.Spec.ClusterIP {
			newSvc.Spec.ClusterIPs = oldSvc.Spec.ClusterIPs
		}
	}
```

if clusterIP was changed:
```
	// clusterIP is not the same
	if oldSvc.Spec.ClusterIP != newSvc.Spec.ClusterIP {
		// this is a client trying to clear it
		if len(oldSvc.Spec.ClusterIP) > 0 && len(newSvc.Spec.ClusterIP) == 0 {
			// if clusterIPs are the same, then clear on their behalf
			if sameStringSlice(oldSvc.Spec.ClusterIPs, newSvc.Spec.ClusterIPs) {
				newSvc.Spec.ClusterIPs = nil
			}
￼			// if they provided nil, then we are fine (handled by patching case above)
￼			// if they changed it then validation will catch it
￼		} else {
￼			// ClusterIP has changed but not cleared *and* ClusterIPs are the same
￼			// then we set ClusterIPs based on ClusterIP
￼			if sameStringSlice(oldSvc.Spec.ClusterIPs, newSvc.Spec.ClusterIPs) {
￼				newSvc.Spec.ClusterIPs = []string{newSvc.Spec.ClusterIP}
￼			}
￼		}
```

If both were changed we leave it alone and let it fail later